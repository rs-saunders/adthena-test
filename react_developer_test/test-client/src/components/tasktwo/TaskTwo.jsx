import React, { useState } from 'react';
import {
  Formik,
  Form,
  Field,
  ErrorMessage,
} from 'formik';
import useUser from './hooks/useUser';
import useTodos from './hooks/useTodos';

const TaskTwo = () => {
  const [currentUsername, setUsername] = useState('Bret');
  const user = useUser(currentUsername);
  const todos = useTodos(user.id);
  const isLoading = todos === null;
  return (
    <div className="task">
      <h1>Task Two</h1>
      <div className="content">
        <Formik
          initialValues={{ username: currentUsername }}
          onSubmit={({ username }) => setUsername(username)}
        >
          {() => (
            <Form>
              <label htmlFor="username">Username:</label>
              <Field
                id="username"
                type="text"
                name="username"
              />
              <ErrorMessage
                name="username"
                component="div"
              />
              <button
                type="submit"
                disabled={isLoading}
              >
                Submit
              </button>
            </Form>
          )}
        </Formik>
        {todos === null
          ? <div>Loading...</div>
          : (
            <>
              <section>
                <h2>{user.name}</h2>
                <dl>
                  <dt>Username</dt>
                  <dd>{user.username}</dd>
                  <dt>Email</dt>
                  <dd>{user.email}</dd>
                  <dt>Website</dt>
                  <dd>{user.website}</dd>
                </dl>
              </section>
              <section>
                <h2>Todo</h2>
                <ul>
                  {todos
                    .filter((todo) => !todo.completed)
                    .map((todo) => <li key={todo.id}>{todo.title}</li>)}
                </ul>
              </section>
              <section>
                <h2>Completed</h2>
                <ul>
                  {todos
                    .filter((todo) => todo.completed)
                    .map((todo) => <li key={todo.id}>{todo.title}</li>)}
                </ul>
              </section>
            </>
          )}
      </div>
    </div>
  );
};

export default TaskTwo;
