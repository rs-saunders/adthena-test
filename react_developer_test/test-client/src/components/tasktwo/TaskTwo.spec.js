import React from 'react';
import { act } from 'react-dom/test-utils';
import { mount } from 'enzyme';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import TaskTwo from './TaskTwo';

describe('TaskTwo', () => {
  it('renders a Loading... message', async (done) => {
    const mock = new MockAdapter(axios);
    mock.onGet('https://jsonplaceholder.typicode.com/todos?userId=1').reply(200, []);

    let wrapper;

    await act(async () => {
      wrapper = mount(<TaskTwo />);
    });

    expect(wrapper).toMatchSnapshot();
    done();
  });

  it('renders a todo list', async (done) => {
    const mock = new MockAdapter(axios);
    mock.onGet('https://jsonplaceholder.typicode.com/todos?userId=1').reply(200, [
      {
        userId: 1,
        id: 1,
        title: 'a task still to do',
        completed: false,
      },
      {
        userId: 1,
        id: 2,
        title: 'a completed task',
        completed: true
      },
    ]);

    let wrapper;

    await act(async () => {
      wrapper = mount(<TaskTwo />);
    });

    wrapper.update();
    expect(wrapper).toMatchSnapshot();
    done();
  });
});
