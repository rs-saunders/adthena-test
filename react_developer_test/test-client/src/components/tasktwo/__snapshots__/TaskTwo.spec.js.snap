// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskTwo renders a Loading... message 1`] = `
<TaskTwo>
  <div
    className="task"
  >
    <h1>
      Task Two
    </h1>
    <div
      className="content"
    >
      <Formik
        initialValues={
          Object {
            "username": "Bret",
          }
        }
        onSubmit={[Function]}
      >
        <Form>
          <form
            action="#"
            onReset={[Function]}
            onSubmit={[Function]}
          >
            <label
              htmlFor="username"
            >
              Username:
            </label>
            <Field
              id="username"
              name="username"
              type="text"
            >
              <input
                id="username"
                name="username"
                onBlur={[Function]}
                onChange={[Function]}
                type="text"
                value="Bret"
              />
            </Field>
            <FormikConnect(ErrorMessageImpl)
              component="div"
              name="username"
            >
              <ErrorMessageImpl
                component="div"
                formik={
                  Object {
                    "dirty": false,
                    "errors": Object {},
                    "getFieldHelpers": [Function],
                    "getFieldMeta": [Function],
                    "getFieldProps": [Function],
                    "handleBlur": [Function],
                    "handleChange": [Function],
                    "handleReset": [Function],
                    "handleSubmit": [Function],
                    "initialErrors": Object {},
                    "initialStatus": undefined,
                    "initialTouched": Object {},
                    "initialValues": Object {
                      "username": "Bret",
                    },
                    "isSubmitting": false,
                    "isValid": true,
                    "isValidating": false,
                    "registerField": [Function],
                    "resetForm": [Function],
                    "setErrors": [Function],
                    "setFieldError": [Function],
                    "setFieldTouched": [Function],
                    "setFieldValue": [Function],
                    "setFormikState": [Function],
                    "setStatus": [Function],
                    "setSubmitting": [Function],
                    "setTouched": [Function],
                    "setValues": [Function],
                    "status": undefined,
                    "submitCount": 0,
                    "submitForm": [Function],
                    "touched": Object {},
                    "unregisterField": [Function],
                    "validateField": [Function],
                    "validateForm": [Function],
                    "validateOnBlur": true,
                    "validateOnChange": true,
                    "validateOnMount": false,
                    "values": Object {
                      "username": "Bret",
                    },
                  }
                }
                name="username"
              />
            </FormikConnect(ErrorMessageImpl)>
            <button
              disabled={true}
              type="submit"
            >
              Submit
            </button>
          </form>
        </Form>
      </Formik>
      <div>
        Loading...
      </div>
    </div>
  </div>
</TaskTwo>
`;

exports[`TaskTwo renders a todo list 1`] = `
<TaskTwo>
  <div
    className="task"
  >
    <h1>
      Task Two
    </h1>
    <div
      className="content"
    >
      <Formik
        initialValues={
          Object {
            "username": "Bret",
          }
        }
        onSubmit={[Function]}
      >
        <Form>
          <form
            action="#"
            onReset={[Function]}
            onSubmit={[Function]}
          >
            <label
              htmlFor="username"
            >
              Username:
            </label>
            <Field
              id="username"
              name="username"
              type="text"
            >
              <input
                id="username"
                name="username"
                onBlur={[Function]}
                onChange={[Function]}
                type="text"
                value="Bret"
              />
            </Field>
            <FormikConnect(ErrorMessageImpl)
              component="div"
              name="username"
            >
              <ErrorMessageImpl
                component="div"
                formik={
                  Object {
                    "dirty": false,
                    "errors": Object {},
                    "getFieldHelpers": [Function],
                    "getFieldMeta": [Function],
                    "getFieldProps": [Function],
                    "handleBlur": [Function],
                    "handleChange": [Function],
                    "handleReset": [Function],
                    "handleSubmit": [Function],
                    "initialErrors": Object {},
                    "initialStatus": undefined,
                    "initialTouched": Object {},
                    "initialValues": Object {
                      "username": "Bret",
                    },
                    "isSubmitting": false,
                    "isValid": true,
                    "isValidating": false,
                    "registerField": [Function],
                    "resetForm": [Function],
                    "setErrors": [Function],
                    "setFieldError": [Function],
                    "setFieldTouched": [Function],
                    "setFieldValue": [Function],
                    "setFormikState": [Function],
                    "setStatus": [Function],
                    "setSubmitting": [Function],
                    "setTouched": [Function],
                    "setValues": [Function],
                    "status": undefined,
                    "submitCount": 0,
                    "submitForm": [Function],
                    "touched": Object {},
                    "unregisterField": [Function],
                    "validateField": [Function],
                    "validateForm": [Function],
                    "validateOnBlur": true,
                    "validateOnChange": true,
                    "validateOnMount": false,
                    "values": Object {
                      "username": "Bret",
                    },
                  }
                }
                name="username"
              />
            </FormikConnect(ErrorMessageImpl)>
            <button
              disabled={false}
              type="submit"
            >
              Submit
            </button>
          </form>
        </Form>
      </Formik>
      <section>
        <h2>
          Bret
        </h2>
        <dl>
          <dt>
            Username
          </dt>
          <dd>
            Leanne Graham
          </dd>
          <dt>
            Email
          </dt>
          <dd>
            Sincere@april.biz
          </dd>
          <dt>
            Website
          </dt>
          <dd>
            hildegard.org
          </dd>
        </dl>
      </section>
      <section>
        <h2>
          Todo
        </h2>
        <ul>
          <li
            key="1"
          >
            a task still to do
          </li>
        </ul>
      </section>
      <section>
        <h2>
          Completed
        </h2>
        <ul>
          <li
            key="2"
          >
            a completed task
          </li>
        </ul>
      </section>
    </div>
  </div>
</TaskTwo>
`;
